### ISSUE TITLE
`Mypy result mismatch for ...`

### ISSUE THREAD

**Issue by Kyle-sandeman-mrdfood** [szabope/mypy-pycharm-plugin#84]
Hello again.
I noticed mypy editor annotations missing from some files but the tool window was reporting issues as usual 

<details><summary>Traceback upon starting pycharm with an offending file open:</summary>
<p>

```
2025-05-07 17:27:04,723 [  14383] SEVERE - #c.i.c.d.i.ExternalToolPass - IdeaLoggingEvent[message=ExternalToolPass: , throwable=com.intellij.diagnostic.PluginException: annotator: works.szabope.plugins.mypy.annotator.MypyAnnotator@17f15ca6 (class works.szabope.plugins.mypy.annotator.MypyAnnotator) [Plugin: works.szabope.mypy]
	at com.intellij.diagnostic.PluginProblemReporterImpl.createPluginExceptionByClass(PluginProblemReporterImpl.java:23)
	at com.intellij.diagnostic.PluginException.createByClass(PluginException.java:90)
	at com.intellij.codeInsight.daemon.impl.ExternalToolPass.processError(ExternalToolPass.java:272)
	at com.intellij.codeInsight.daemon.impl.ExternalToolPass.doApply(ExternalToolPass.java:245)
	at com.intellij.codeInsight.daemon.impl.ExternalToolPass.doApply(ExternalToolPass.java:235)
	at com.intellij.codeInsight.daemon.impl.ExternalToolPass$1.lambda$run$1(ExternalToolPass.java:178)
	at com.intellij.openapi.application.ReadAction.lambda$run$1(ReadAction.java:54)
	at com.intellij.openapi.application.impl.AnyThreadWriteThreadingSupport.runReadAction(AnyThreadWriteThreadingSupport.kt:314)
	at com.intellij.openapi.application.impl.AnyThreadWriteThreadingSupport.runReadAction(AnyThreadWriteThreadingSupport.kt:262)
	at com.intellij.openapi.application.impl.ApplicationImpl.runReadAction(ApplicationImpl.java:864)
	at com.intellij.openapi.application.ReadAction.compute(ReadAction.java:66)
	at com.intellij.openapi.application.ReadAction.run(ReadAction.java:53)
	at com.intellij.codeInsight.daemon.impl.ExternalToolPass$1.lambda$run$2(ExternalToolPass.java:175)
	at com.intellij.openapi.progress.util.BackgroundTaskUtil.lambda$runUnderDisposeAwareIndicator$15(BackgroundTaskUtil.java:371)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcess$1(CoreProgressManager.java:223)
	at com.intellij.platform.diagnostic.telemetry.helpers.TraceKt.use(trace.kt:45)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcess$2(CoreProgressManager.java:222)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$executeProcessUnderProgress$14(CoreProgressManager.java:674)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:749)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computeUnderProgress(CoreProgressManager.java:705)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:673)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:79)
	at com.intellij.openapi.progress.impl.CoreProgressManager.runProcess(CoreProgressManager.java:203)
	at com.intellij.openapi.progress.util.BackgroundTaskUtil.runUnderDisposeAwareIndicator(BackgroundTaskUtil.java:366)
	at com.intellij.codeInsight.daemon.impl.ExternalToolPass$1.run(ExternalToolPass.java:172)
	at com.intellij.util.ui.update.Update.execute$suspendImpl(Update.kt:61)
	at com.intellij.util.ui.update.Update.execute(Update.kt)
	at com.intellij.util.ui.update.MergingUpdateQueue.executeUpdates$suspendImpl(MergingUpdateQueue.kt:371)
	at com.intellij.util.ui.update.MergingUpdateQueue.executeUpdates(MergingUpdateQueue.kt)
	at com.intellij.util.ui.update.MergingUpdateQueue$restart$1.invokeSuspend(MergingUpdateQueue.kt:351)
	at com.intellij.util.ui.update.MergingUpdateQueue$restart$1.invoke(MergingUpdateQueue.kt)
	at com.intellij.util.ui.update.MergingUpdateQueue$restart$1.invoke(MergingUpdateQueue.kt)
	at com.intellij.util.SingleAlarm$scheduleTask$1$1$1.invokeSuspend(SingleAlarm.kt:394)
	at com.intellij.util.SingleAlarm$scheduleTask$1$1$1.invoke(SingleAlarm.kt)
	at com.intellij.util.SingleAlarm$scheduleTask$1$1$1.invoke(SingleAlarm.kt)
	at kotlinx.coroutines.intrinsics.UndispatchedKt.startUndispatchedOrReturn(Undispatched.kt:62)
	at kotlinx.coroutines.BuildersKt__Builders_commonKt.withContext(Builders.common.kt:155)
	at kotlinx.coroutines.BuildersKt.withContext(Unknown Source)
	at com.intellij.util.SingleAlarm$scheduleTask$1$1.invokeSuspend(SingleAlarm.kt:392)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:104)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:608)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:873)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:763)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:750)
Caused by: java.lang.IllegalArgumentException: Mypy result mismatch for MypyOutput(file=/<redacted>_pb2.pyi, line=295, column=4, message=Need type annotation for "__slots__" (hint: "__slots__: list[<type>] = ..."), hint=null, code=var-annotated, severity=ERROR)
	at works.szabope.plugins.mypy.annotator.MypyAnnotator.apply(MypyAnnotator.kt:49)
	at works.szabope.plugins.mypy.annotator.MypyAnnotator.apply(MypyAnnotator.kt:21)
	at com.intellij.codeInsight.daemon.impl.AnnotationHolderImpl.applyExternalAnnotatorWithContext(AnnotationHolderImpl.java:240)
	at com.intellij.codeInsight.daemon.impl.ExternalToolPass.doApply(ExternalToolPass.java:242)
	... 41 more
]
```

Not sure if the "41 more" contains anything helpful, but if it does I am unsure how to make pycharm spit it out.
</p>
</details> 

I'm not having any success in creating a minimal reproducible example for you, so we may need a bit of back-and-forth to get to the bottom of it.
What I can say is that the file in the output above is a Google Protobuf (compiled stub). This is imported as a transitive dependency of the file in my editor.
This may help

```python
#myfile.py
from somepackage import helper
from another_thing import foobar
x: int = '1'
```
```python
# helper.py
from path.to.compiled_protobufs.xyz_pb2 import abc
...
```
```python
# xyz_pb2.pyi
...
class Status(int, metaclass=_enum_type_wrapper.EnumTypeWrapper):
    __slots__ = []

```

If I comment out the import in `myfile` then the plugin correctly highlights `'1'` as a mypy failure.
If I run mypy via the CLI, I still get all the usual output (even if I use the mypy command your plugin is using, except the shadow file).
Here's an example of the mypy invocation, but NB the traceback I showed above is NOT printed more than once per pycharm session, I think.
```sh
2025-05-07 17:49:35,104 [1364764]   FINE - #works.szabope.plugins.mypy.services.MypyService - Executing command: /path/to/mypy --show-column-numbers --show-absolute-path --output json --follow-imports silent --exclude '\.pyi$' --shadow-file <workdir>/myfile.py /var/folders/g8/sz1j3bqd2fn6dtt84x9pp0mh0000gq/T/pycharm_mypy_13948425607082515614_myfile.py <workdir>/myfile.py with workDir: <workdir>
```
I cannot tell the difference in your TRACE logs between a successful run and an unsuccessful one. I suspect the issue is in the parsing of the mypy output, so it does not seem to me like mypy is producing bad output.  
There are quite a lot of `__slots__` related complaints assumedly because of the way google generates protobuf files.  
By comparison, commenting out `foobar`'s import has no effect if `foobar` does not result in importing a protobuf.

*Important*:
The manual scan returns the correct output (like CLI) even when the editor highlights are not working.
However it seems like I have to sometimes delete the cache for it to actually see my latest change.

**Comment by Kyle-sandeman-mrdfood** [comment]
To add to this, in some cases I get the `__slots__` annotation appearing in the current file while it comes from a Protobuf stub file.
e.g.
```
tests/
  test_stuff.py
main_dir/
  compiled_protobufs/
    something_pb2.pyi
```
The file in the editor here is test_stuff and one can easily see there's nothing to do with slots (it is just a dictionary literal)
I can only assume that the plugin is not considering the filename in mypy's output when generating the in-editor errors.
![Image](https://github.com/user-attachments/assets/93e4bddd-5d08-4433-88fe-812a128a2c1d)

**Comment by Kyle-sandeman-mrdfood** [comment]
I have done this to exclude these errors from google protobuff files for now
`config.ini`
```ini
[mypy-*.compiled_protobuffs.*]
disable_error_code = var-annotated
```

**Comment by lattwood** [comment]
I'm also reporting this issue.

**Comment by szabope** [comment]
Hey @Kyle-sandeman-mrdfood 

I really want to figure this out, because I just can't imagine how it is possible.
Thank you for your explanation. I'll try again to reproduce it.

**Comment by Kyle-sandeman-mrdfood** [comment]
That's great to hear, thank you!
